File: c_files/raylib/bouncing_balls.c
#include <raylib.h>
#include <raymath.h>

#define NUM_OF_BALLS 200

// BALL
typedef struct Ball
{
  float x;
  float y;
  float radius;
  float speed;
  Vector2 direction;
  Color color;

} Ball;

void
ball_init(Ball* ball)
{
  ball->radius = GetRandomValue(10, 30);
  ball->x = GetRandomValue(ball->radius, GetScreenWidth() - ball->radius);
  ball->y = GetRandomValue(ball->radius, GetScreenHeight() - ball->radius);
  ball->speed = GetRandomValue(100, 300);
  ball->direction = (Vector2){
    GetRandomValue(0, 1) == 0 ? -1 : 1,
    GetRandomValue(0, 1) == 0 ? -1 : 1,
  };
  ball->color = (Color){
    GetRandomValue(0, 255), GetRandomValue(0, 255), GetRandomValue(0, 255), 255
  };
}

void
ball_draw(Ball* ball)
{
  DrawCircleV((Vector2){ ball->x, ball->y }, ball->radius, ball->color);
}

void
ball_update(Ball* ball)
{
  // move ball
  ball->x += ball->direction.x * ball->speed * GetFrameTime();
  ball->y += ball->direction.y * ball->speed * GetFrameTime();

  // normalize direction vector
  if (ball->direction.x != 0 && ball->direction.y != 0) {
    ball->direction = Vector2Normalize(ball->direction);
  }

  // ball bounds
  if (ball->x < ball->radius || ball->x > GetScreenWidth() - ball->radius) {
    ball->direction.x *= -1;
  }
  if (ball->y < ball->radius || ball->y > GetScreenHeight() - ball->radius) {
    ball->direction.y *= -1;
  }
}

// MAIN
int
main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Bouncing Balls";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  // INIT
  Ball balls[NUM_OF_BALLS];

  for (int i = 0; i < NUM_OF_BALLS; i++) {
    Ball ball;
    ball_init(&ball);

    balls[i] = ball;
  }

  while (!WindowShouldClose()) {

    // UPDATE
    for (int i = 0; i < NUM_OF_BALLS; i++) {
      ball_update(&balls[i]);
    }

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    // DRAW
    for (int i = 0; i < NUM_OF_BALLS; i++) {
      ball_draw(&balls[i]);
    }

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python_files/pygame/bouncing_balls.py
import random
import sys

import pygame as p

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 800
SCREEN_TITLE: str = "Bouncing Balls"
SCREEN_BACKGROUND: str = "white"
FPS: int = 60


class Ball:
    def __init__(self) -> None:
        self.radius: float = random.uniform(5, 30)
        self.center: p.math.Vector2 = p.math.Vector2(
            random.uniform(self.radius, SCREEN_WIDTH - self.radius),
            random.uniform(self.radius, SCREEN_HEIGHT - self.radius),
        )
        self.speed: float = random.uniform(100, 300)
        self.direction: p.math.Vector2 = p.math.Vector2(
            -1 if random.randint(0, 1) == 0 else 1,
            -1 if random.randint(0, 1) == 0 else 1,
        )
        self.color: p.Color = p.Color(
            random.randint(0, 255),
            random.randint(0, 255),
            random.randint(0, 255),
        )

    def draw(self, surface: p.Surface) -> None:
        p.draw.circle(surface, self.color, self.center, self.radius)

    def update(self, dt: float) -> None:
        # move
        self.center.x += self.direction.x  * self.speed * dt
        self.center.y += self.direction.y  * self.speed * dt

        # bounds
        if self.center.x < self.radius || self.center.x > SCREEN_WIDTH - self.radius:
            self.direction.x *= -1
        if self.center.y < self.radius || self.center.y > SCREEN_HEIGHT - self.radius:
            self.direction.y *= -1


class Game:
    def __init__(self) -> None:
        p.init()

        self.screen: p.Surface = p.display.set_mode(
            (SCREEN_WIDTH, SCREEN_HEIGHT)
        )
        p.display.set_caption(SCREEN_TITLE)

        self.clock: p.time.Clock = p.time.Clock()
        self.running: bool = False

    def handle_events(self) -> None:
        for event in p.event.get():
            if event.type == p.QUIT:
                self.running = False

    def draw(self) -> None:
        self.screen.fill(SCREEN_BACKGROUND)

        p.display.flip()

    def update(self, dt: float) -> None:
        pass

    def run(self) -> None:
        while self.running:
            # delta time
            dt: float = self.clock.tick(FPS) / 1000.0

            self.handle_events()
            self.update(dt)
            self.draw()

        p.quit()
        sys.exit()
